Loop/Managers/DeviceAlert/InAppModalDeviceAlertHandler.swift

```diff
@@ -0,0 +1,141 @@
+                if !repeats {
+                    self?.clearPendingAlert(identifier: alert.identifier)
+                }
+            }
+            self.addPendingAlert(alert: alert, timer: timer)
+        }
+    }
+    
+    private func show(alert: DeviceAlert) {
+        guard let content = alert.foregroundContent else {
+            return
+        }
+        DispatchQueue.main.async {
+            if self.isAlertShowing(identifier: alert.identifier) {
+                return
+            }
+            let alertController = self.presentAlert(title: content.title, message: content.body, action: content.acknowledgeActionButtonLabel) { [weak self] in
+                self?.clearDeliveredAlert(identifier: alert.identifier)
+                self?.deviceAlertManagerResponder?.acknowledgeDeviceAlert(identifier: alert.identifier)
+            }
+            self.addDeliveredAlert(alert: alert, controller: alertController)
+        }
+    }
+    
+    private func addPendingAlert(alert: DeviceAlert, timer: Timer) {
+        dispatchPrecondition(condition: .onQueue(.main))
+        self.alertsPending[alert.identifier] = (timer, alert)
+    }
+
+    private func addDeliveredAlert(alert: DeviceAlert, controller: UIAlertController) {
+        dispatchPrecondition(condition: .onQueue(.main))
+        self.alertsShowing[alert.identifier] = (controller, alert)
+    }
+    
+    private func clearPendingAlert(identifier: DeviceAlert.Identifier) {
+        dispatchPrecondition(condition: .onQueue(.main))
+        alertsPending[identifier] = nil
+    }
+
+    private func clearDeliveredAlert(identifier: DeviceAlert.Identifier) {
+        dispatchPrecondition(condition: .onQueue(.main))
+        alertsShowing[identifier] = nil
+    }
+    
+    private func isAlertPending(identifier: DeviceAlert.Identifier) -> Bool {
+        dispatchPrecondition(condition: .onQueue(.main))
+        return alertsPending.index(forKey: identifier) != nil
+    }
+    
+    private func isAlertShowing(identifier: DeviceAlert.Identifier) -> Bool {
+        dispatchPrecondition(condition: .onQueue(.main))
+        return alertsShowing.index(forKey: identifier) != nil
+    }
+
+    private func presentAlert(title: String, message: String, action: String, completion: @escaping () -> Void) -> UIAlertController {
+        dispatchPrecondition(condition: .onQueue(.main))
+        // For now, this is a simple alert with an "OK" button
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        alertController.addAction(UIAlertAction(title: action, style: .default, handler: { _ in completion() }))
+        topViewController(controller: rootViewController)?.present(alertController, animated: true)
+        return alertController
+    }
+    
+    // Helper function pulled from SO...may be outdated, especially in the SwiftUI world
+    private func topViewController(controller: UIViewController?) -> UIViewController? {
```

Comment:
Did you verify that this was needed? I.e. that presenting from the rootViewController directly results in an obscured alert?

Reply:
I'm not sure in every case in our app, and I certainly am not sure about every SwiftUI screen. This was a suggestion from StackOverflow.